### DISCUSSION TITLE
Best way to add business logic to Entity

### DISCUSSION THREAD

**Discussion by baznikin** [noam-honig/radweb#539] [Q&A] [answered]
Hello!

I am stating a project, never code on TS before and thinking how to organize code best way. At time I want to update other exemplar of same entity on save:

```
//      accepted - collection owner selected ID
//      outdated - previous accepted ID (for historical reasons)
const idstatuses = ['accepted', 'outdated', 'couple', 'more', 'statuses' ] as const
export type IdentificationStatus = (typeof idstatuses)[number]

@Entity('identifications', {
  allowApiCrud: true,
})
export class Identification {
  @Fields.cuid({ allowApiUpdate: false  })
  id!: string

  @Relations.toOne(() => Specimen)
  specimen?: Specimen

  @Fields.literal(() => idstatuses, {
    caption: 'Status',
  })
  status?: IdentificationStatus
}

export class Specimen {
  @Fields.cuid({ allowApiUpdate: false  })
  id!: string

  @Relations.toMany(() => Identification)
  identifications?: Identification[]
}
```

Upon saving of entity Identification I want to check if its new status is `accepted`. If it is `accepted`, then find between Identifications of linked Specimen one with status `accepted` and change it to `outdated`.
1. My first fought was lifecycle hook `saving` or `saved`. Is it appropriate place for such logic?
2. Should I move actual code to other module to keep entity code clean? Where is better to put it? Should it be BackendMethod?

**Comment by noam-honig** [comment, marked as answer]
Hi @baznikin, I would put that code in the saving/saved lifecycle hook - I wouldn't put it in a backend method - since you mentioned that you wanted this logic to happen always - and not just when the backend method is called.

As for organization, I would start by putting it in the hook itself - and if / when I'll feel that it's becoming harder to follow / maintain I'll extract it to a function in the same file - and eventually if that's too big move it to another file.

the code would be something like:
```ts
@Entity<Identification >('identifications', {
  allowApiCrud: true,
  saved:async(id,e)=>{
    if (e.fields.status.valueChanged()&&id.status=='accepted'){
      ....
    }
  }
})
```

**Comment by baznikin** [comment]
Thanks for snippet!
Another question I wanted to ask - how do you cover business logic in entities with tests? What framework/approach used? For sure, no reason to check functionality provided by remult itself, but I want to make sure I'll not broke my own business logic somehow

**Comment by noam-honig** [comment]
Hi @baznikin, 
One of the fun parts about remult is that you can easily abstract the data access layer and run tests on that.

Checkout this test example from one of my projects:
https://github.com/noam-honig/app-with-signup-and-basic-user-management/blob/master/src/components/user/auth-controller.spec.ts

See how I set the `dataProvider` to an `inMemory` data provider and run the tests with it

**Reply by baznikin** [reply]
Thank you! `<3`

**Comment by baznikin** [comment]
Feeling bad for asking newbie questions, but I can't make it work. First of all, I tried this approach to find records to update, but found `specimen` relation not loaded:

```
idt.specimen?.identifications?.filter((x) => x != idt && x.status == 'accepted')
```
After lots of hacky tries best I get is this one (also I convert Entity to Active Record form, anyway I wanted to do it after). How to do it properly?

```
  saved: async (idt, e) => {
    if (e.fields.status.valueChanged()) {
      console.log('STATUS CHANGED ', idt)
    }
    if (e.fields.status.valueChanged() && idt.status == 'accepted') {
      const specimenRepo = remult.repo(Specimen)
      const idRepo = remult.repo(Identification)
      const thisId = await idRepo.findFirst(idt)
      if (!thisId) { return }
      const spcmn = await idRepo.relations(thisId).specimen.findOne()
      if (!spcmn) { return }
      console.log('SPECIMEN ',spcmn)
      const otherIds = await specimenRepo.relations(spcmn).identifications.find({where: {status: 'accepted', $not: { id: idt.id} }})
      console.log('OTHER ACCEPTED IDS ',otherIds)
      otherIds.forEach((x) => { x.status='outdated'; x.save}) // do not saved
      otherIds.forEach((x) => { x.status='outdated'; idRepo.update(x)}) // idRepo.update - An argument for 'item' was not provided.

      // const prevAccepted = idt.specimen?.identifications?.filter((x) => x != idt && x.status == 'accepted')
      // for(const item in prevAccepted){
      //   console.log('saved ', item)
      // }
      
      // prevAccepted?.forEach(function (itd) { idt?.status = 'outdated' && idt.save! })
    }
  },
})
export class Identification extends EntityBase {
```

**Comment by baznikin** [comment]
Best I can get is this. Not sure is Admin UI uses LiveView, it do not reflect changes to other records till reload. What can be improved?
```
    if (e.fields.status.valueChanged() && idt.status == 'accepted') {
      const idRepo = remult.repo(Identification)
      const thisId = await idRepo.findFirst(idt)

      if (!thisId) {
        return
      }
      const spcmn = await idRepo.relations(thisId).specimen.findOne({include: {identifications: true} })

      if (!spcmn || !spcmn.identifications) {
        return
      }

      for (const oid of spcmn.identifications.filter((x) => x != idt && x.status == 'accepted')) {
        await idRepo.update(oid, { status: 'outdated' })
      }
    }
  }
```

Update - a little bit better, stil not ideal :D Can't get rid of `if + return` and `oid.save` still not working
```
    if (e.fields.status.valueChanged() && idt.status == 'accepted') {
      const idRepo = remult.repo(Identification)

      const spcmn = await idt._.relations.specimen.findOne({
        include: {
          identifications: {
            where: {
              status: 'accepted',
              $not: { id: idt.id },
            },
          },
        },
      })

      if (!spcmn || !spcmn.identifications) {
        return
      }

      for (const oid of spcmn.identifications) {
        console.log('REVERSING ', oid)
        // oid.status = 'outdated'
        // oid.save
        await idRepo.update(oid, { status: 'outdated' })
      }
    }
  }
```

**Comment by noam-honig** [comment]
Not sure that I fully grasp you example so I created a test project for it on stackblitz.
https://stackblitz.com/edit/stackblitz-starters-1voywg?file=test.spec.ts

Here's the test:
```ts
describe('The tests', async () => {
  let s1: Specimen;
  beforeEach(async () => {
    remult.dataProvider = new InMemoryDataProvider();
    s1 = await repo(Specimen).insert({});
    await repo(Specimen)
      .relations(s1)
      .identifications.insert({ status: 'accepted' });
  });
  test('New Identification updates previous one as out dated', async () => {
    let currentId = await repo(Identification).findFirst();
    await repo(Identification).insert({
      specimen: s1,
      status: 'accepted',
    });
    currentId = await repo(Identification).findId(currentId.id);

    expect(currentId.status).toBe('outdated');
  });
});
```

Here's the `saving` code that I got to:
```ts
@Entity<Identification>('identifications', {
  allowApiCrud: true,
  saving: async (id, e) => {
    if (
      (e.fields.status.valueChanged() || e.isNew) &&
      id.status == 'accepted'
    ) {
      const s = await e.fields.specimen.load();

      await repo(Specimen)
        .relations(s)
        .identifications.updateMany({
          where: {
            specimen: s,
            id: { $not: id.id },
            status: 'accepted',
          },
          set: {
            status: 'outdated',
          },
        });
    }
  },
})
```

I'm using the `e.fields.speciment.load()` to make sure the relation is loaded.
Another alternative is to store also the `specimentId` in a separate field - that'll remove the need of loading the speciment just for using it in the filter.
https://remult.dev/docs/entity-relations#custom-relation-field


Tryout the stackblitz project, and adjust the test as needed

p.s. I'll be leaving for a holiday till Sunday later today, so I appologize if I'll only respond then

**Comment by baznikin** [comment]
Woah, so many good points to consider! Many thanks, @noam-honig !
